### PR TITLE
Avoid Memo.Sync

### PR DESCRIPTION
**Pull request by snowleopard**
@jeremiedimino Removing some of the remaining uses of `Memo.Sync` led to another huge refactoring. I decided to pause mid-way to discuss whether to proceed with it or not.

P.S.: I've seen that your draft PR #4343 on the same topic but it looks like you started by first working out some abstractions, so hopefully we didn't duplicate a lot of work.

### CODE REVIEW

File: src/dune_rules/expander.ml [hunk 1]
```diff
@@ -306,10 +307,14 @@ let expand_pform_gen ~(context : Context.t) ~bindings ~dir ~source
       | Project_root ->
         Need_full_expander
           (fun t ->
-            Action_builder.return
-              [ Value.Dir (Path.build (Scope.root t.scope)) ])
-      | Cc -> Need_full_expander (fun t -> (cc t).c)
-      | Cxx -> Need_full_expander (fun t -> (cc t).cxx)
+            Memo.Build.return
```
Comment (jeremiedimino) [inline src/dune_rules/expander.ml:310]: Something doesn't feel quite right here. Morally, an `'a Action_builder.t` is a `('a * Dep.Facts.t) Memo.Build.t`, so a `'a Action_builder.t Memo.Build.t` is a nested monad. I had the same issue in #4343 and solved it by "injecting" my `Memo.Build.t` values into `Action_builder.t` via `Action_builder.memo_build`. Could we do the same here?

**Review by jeremiedimino — CHANGES_REQUESTED** [review]

File: src/dune_rules/fdo.ml [hunk 2]
```diff
@@ -91,38 +91,39 @@ let get_profile =
...
-            [ Pp.textf "%s=%s but profile file %s does not exist." Mode.var
-                (Mode.to_string Always) (Path.to_string path)
-            ]
-      | Never -> false
-    in
-    if use_profile then
-      Some path
-    else
-      None
+    Memo.Build.return
```
Comment (nojb) [inline src/dune_rules/fdo.ml:94]: This is a trivial remark, but the diff would be smaller if you push the call to `Memo.Build.return` to the leaves of the expression (`Some path` and `None`). And also easier to read IMHO due to the reduced nesting. I think there are a few other places where the same remark could apply.

Comment (snowleopard) [reply src/dune_rules/fdo.ml:94]: Agreed! Fixed here, and will look out for similar cases.

File: src/dune_rules/expander.ml (hunk 1 above)
Comment (snowleopard) [reply src/dune_rules/expander.ml:310]: Indeed, I already had to use `Action_builder.memo_build` in a couple of places. I'll see if I can use it here too.

Comment (snowleopard) [reply src/dune_rules/expander.ml:310]: Oh, wow, that seems to simplify a lot. Partially undoing my changes :joy:

**Comment by snowleopard** [comment]
OK, I made it work but I'd like to make another pass and switch from sequential to parallel traversals where possible.

File: bin/printenv.ml [hunk 3]
```diff
@@ -59,15 +60,19 @@ let term =
             let sctx =
               Dune_engine.Context_name.Map.find_exn setup.scontexts ctx.name
             in
-            [ dump sctx ~dir:(Path.as_in_build_dir_exn dir) ]
+            [ Action_builder.memo_build
+                (dump sctx ~dir:(Path.as_in_build_dir_exn dir))
+              |> Action_builder.bind ~f:Fun.id
```
Comment (nojb) [inline bin/printenv.ml:65]: Can use `memo_build_join` ?

File: bin/printenv.ml [hunk 4]
```diff
@@ -59,15 +60,19 @@ let term =
...
           | In_source_dir dir ->
             Dune_engine.Context_name.Map.values setup.scontexts
             |> List.map ~f:(fun sctx ->
                    let dir =
                      Path.Build.append_source
                        (Super_context.context sctx).build_dir dir
                    in
-                   dump sctx ~dir)
+                   Action_builder.memo_build (dump sctx ~dir)
+                   |> Action_builder.bind ~f:Fun.id)
```
Comment (nojb) [inline bin/printenv.ml:75]: Use `memo_build_join` ?

File: src/dune_rules/super_context.ml [hunk 5]
```diff
@@ -112,38 +113,46 @@ end = struct
...
-    in
-    let expander =
-      Memo.Lazy.map expander_for_artifacts ~f:(fun expander_for_artifacts ->
-          extend_expander t ~dir ~expander_for_artifacts)
-    in
-    Env_node.make ~dir ~scope ~config_stanza ~inherit_from:(Some inherit_from)
-      ~profile:t.context.profile ~expander ~expander_for_artifacts
-      ~default_context_flags ~default_env:t.context_env
-      ~default_bin_artifacts:t.bin_artifacts
+    Memo.Build.return
```
Comment (nojb) [inline src/dune_rules/super_context.ml:116]: Can push down this call to `Env_node.make` below.

**Comment by nojb** [comment]
Sorry for the drive-by question, but I don't have a feeling for when one should use `'a Memo.Build.t` and when one should use `'a Memo.Lazy.t`. Is there an intuitive idea one can have about this?

File: bin/printenv.ml (hunk 3 above)
Comment (snowleopard) [reply bin/printenv.ml]: Indeed, fixed.

File: bin/printenv.ml (hunk 4 above)
Comment (snowleopard) [reply bin/printenv.ml]: Fixed.

File: src/dune_rules/super_context.ml (hunk 5 above)
Comment (snowleopard) [reply src/dune_rules/super_context.ml]: Thanks, well-spotted!

File: src/dune_rules/fdo.ml [hunk 6]
```diff
@@ -168,21 +175,28 @@ module Linker_script = struct
     let linker_script_path =
       Path.Build.(relative ctx.build_dir (Path.to_string linker_script))
     in
+    let open Memo.Build.O in
     let flags () =
       let open Command.Args in
-      match get_profile ctx with
+      let* () = Memo.Build.return () in
```
Comment (nojb) [inline src/dune_rules/fdo.ml:181]: Just to make sure I understand, this is here to delay the computation of the following expression?

In any case, can't the `()` argument to flags be removed?

File: src/dune_rules/env_node.ml [hunk 7]
```diff
@@ -15,90 +15,97 @@ end
...
     inherited ~field:external_env ~root:default_env (fun env ->
         let env, have_binaries =
           (Env.extend_env env config.env_vars, List.is_non_empty config.binaries)
         in
-        if have_binaries then
-          let dir = Utils.local_bin dir |> Path.build in
-          Env.cons_path env ~dir
-        else
-          env)
+        Memo.Build.return
```
Comment (nojb) [inline src/dune_rules/env_node.ml:98]: Here the difference is small, but you could push down this call as well.

**Comment by jeremiedimino** [comment]
> Sorry for the drive-by question, but I don't have a feeling for when one should use 'a Memo.Build.t and when one should use 'a Memo.Lazy.Async.t. Is there an intuitive idea one can have about this?

Yes, there is a simple mental model: if you use the same `'a Memo.Build.t` value twice, the computation described by this monad will be executed twice. On the other hand, if you call `Memo.Lazy.Async.force` multiple times on the same `Memo.Lazy.Async.t`, the computation will be executed only once and the result will be shared between the multiple calls to `Memo.Lazy.Async.t`. And in some case, it will even be shared across builds in polling mode.

The tradeoff is that a `Memo.Lazy.Async.t` has a bigger memory footprint than a `Memo.Build.t`; the former is a mutable object and it will hold onto the result while the latter is merely a `continuation:('a -> unit) -> unit` function.

**Comment by jeremiedimino** [comment]
I've scheduled myself for a presentation of how to work with these new APIs BTW. I have learned quite a lot in my recent PRs and I'm eager to talk about it :)

**Comment by nojb** [comment]
> I've scheduled myself for a presentation of how to work with these new APIs BTW. I have learned quite a lot in my recent PRs and I'm eager to talk about it :)

That will be great! Thanks!

File: src/dune_rules/env_node.ml (hunk 7 above)
Comment (snowleopard) [reply src/dune_rules/env_node.ml:98]: Done.

File: src/dune_rules/module_compilation.ml [hunk 8]
```diff
@@ -63,182 +62,197 @@ let build_cm cctx ~dep_graphs ~precompiled_cmi ~cm_kind (m : Module.t) ~phase =
...
-  let linear_fdo =
-    Obj_dir.Module.obj_file obj_dir m ~kind:Cmx ~ext:Fdo.linear_fdo_ext
-  in
-  let extra_args, extra_deps, other_targets =
-    if precompiled_cmi then
-      (force_read_cmi src, [], [])
-    else
-      (* If we're compiling an implementation, then the cmi is present *)
-      let public_vlib_module = Module.kind m = Impl_vmodule in
+  Memo.Build.Option.iter ~f:Fun.id
```
Comment (nojb) [inline src/dune_rules/module_compilation.ml:65]: I think here the call could also be pushed further down.

File: src/dune_rules/fdo.ml (hunk 6 above)
Comment (snowleopard) [reply src/dune_rules/fdo.ml:181]: Yes, binding on unit delays the synchronous part of the computation until the resulting `Memo.Build.t` is executed. I just realised that I actually had two such delays, so I deleted one of them.

The unit argument of `flags` is still useful: if we omit it, `get_profile` will always be executed, which I'm guessing is undesirable (I'm not the original author of this code but `Action_builder.delayed` must have been there for a reason).

Comment (nojb) [reply src/dune_rules/fdo.ml]: Good point, makes sense.

File: src/dune_rules/module_compilation.ml (hunk 8 above)
Comment (snowleopard) [reply src/dune_rules/module_compilation.ml:65]: Indeed, I pushed this bit down, where its equivalent was originally.

**Comment by snowleopard** [comment]
Here are results of `perf.sh`:

```
             Current branch Main branch
             ============== ===========
           | real 1.541s    real 1.543s
Full build | user 1.277s    user 1.289s
           | sys  0.224s    sys  0.214s
                            
           | real 0.714s    real 0.758s
Zero build | user 0.603s    user 0.642s
           | sys  0.097s    sys  0.083s
```

No visible performance impact.

File: src/dune_rules/file_binding.mli [hunk 9]
```diff
@@ -21,6 +21,12 @@ module Unexpanded : sig
   val make : src:Loc.t * string -> dst:Loc.t * string -> t
 
   val expand :
+       t
+    -> dir:Path.Build.t
+    -> f:(String_with_vars.t -> string Memo.Build.t)
+    -> Expanded.t Memo.Build.t
+
+  val expand_pure :
```
Comment (jeremiedimino) [inline src/dune_rules/file_binding.mli:29]: We generally use the term "static" in other places. Though, I expect that will eventually get rid of them all once we have pushed the monad enough.

**Review by jeremiedimino — APPROVED** [review]
Looks good

File: src/dune_rules/file_binding.mli (hunk 9 above)
Comment (snowleopard) [reply src/dune_rules/file_binding.mli]: Indeed, fixed.

**Comment by nojb** [comment]
Just to understand, is there a technical obstacle to adapting the remaining uses of `Sync`, eg https://github.com/ocaml/dune/blob/d45c3521b0be75f47d1188b65f999d8fb0d496eb/src/dune_engine/file_tree.ml#L637-L640
?

**Comment by snowleopard** [comment]
@nojb We could in principle tackle everything in one PR but then the PR grows bigger and harder to review. We decided to stage the changes. The remaining occurrences of `Sync` will be taken care of subsequent PRs.

**Comment by nojb** [comment]
> @nojb We could in principle tackle everything in one PR but then the PR grows bigger and harder to review. We decided to stage the changes. The remaining occurrences of `Sync` will be taken care of in subsequent PRs.

OK, that's what I was wondering. Thanks!

**Pull request merged**